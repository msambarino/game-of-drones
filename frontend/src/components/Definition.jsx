import React from 'react'
import PropTypes from 'prop-types'
import { DEFAULT_RULES } from 'data/constants'

class Definition extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      useDefaultRules: true,
      rules: DEFAULT_RULES,
      player1: { value: '', error: null },
      player2: { value: '', error: null }
    }

    this.handlePlayer1Change = this.handlePlayer1Change.bind(this)
    this.handlePlayer2Change = this.handlePlayer2Change.bind(this)
    this.handleStartClick = this.handleStartClick.bind(this)
    this._checkFieldValues = this._checkFieldValues.bind(this)
  }

  handlePlayer1Change(event) {
    const value = event.target.value
    this.setState({
      player1: { value, error: value === '' ? "Can't be blank" : null },
      player2: {
        ...this.state.player2,
        error:
          this.state.player1.value === this.state.player2.value &&
          value !== this.state.player2.value &&
          this.state.player2.value !== ''
            ? null
            : this.state.player2.error
      }
    })
  }

  handlePlayer2Change(event) {
    const value = event.target.value
    this.setState({
      player2: {
        value,
        error:
          value === ''
            ? "Can't be blank"
            : value === this.state.player1.value
              ? "Can't be the same as player 1"
              : null
      }
    })
  }

  handleStartClick() {
    const errors = this._checkFieldValues()
    if (errors.player1 || errors.player2) {
      this.setState({
        player1: {
          ...this.state.player1,
          error: errors.player1
        },
        player2: {
          ...this.state.player2,
          error: errors.player2
        }
      })
    } else {
      this.props.handleGameStart({
        player1: this.state.player1.value,
        player2: this.state.player2.value,
        rules: this.state.rules
      })
    }
  }

  _checkFieldValues() {
    return {
      player1: this.state.player1.value === '' ? "Can't be blank" : null,
      player2:
        this.state.player2.value === ''
          ? "Can't be blank"
          : this.state.player2.value === this.state.player1.value
            ? "Can't be the same as player 1"
            : null
    }
  }

  render() {
    const { player1, player2 } = this.state
    return (
      <div className="container">
        <section className="section is-small columns is-centered">
          <div className="column is-half">
            <h3 className="title has-text-centered">Enter Player's Names</h3>
            <div className="field">
              <label className="label is-normal">Player 1</label>
              <div className="field">
                <div className="control">
                  <input
                    className={`input ${player1.error && 'is-danger'}`}
                    type="text"
                    placeholder="Player 1 name"
                    value={player1.value}
                    onChange={this.handlePlayer1Change}
                  />
                  {player1.error ? (
                    <p className="help is-danger">{player1.error}</p>
                  ) : null}
                </div>
              </div>
            </div>
            <div className="field">
              <label className=" label is-normal">Player 2</label>
              <div className="field">
                <div className="control">
                  <input
                    className={`input ${player2.error && 'is-danger'}`}
                    type="text"
                    placeholder="Player 2 name"
                    value={player2.value}
                    onChange={this.handlePlayer2Change}
                  />
                  {player2.error ? (
                    <p className="help is-danger">{player2.error}</p>
                  ) : null}
                </div>
              </div>
            </div>
            <div className="control">
              <button
                className="button is-primary is-fullwidth"
                onClick={this.handleStartClick}
              >
                Start
              </button>
            </div>
          </div>
        </section>
      </div>
    )
  }
}

Definition.propTypes = {
  handleGameStart: PropTypes.func.isRequired
}

export default Definition
