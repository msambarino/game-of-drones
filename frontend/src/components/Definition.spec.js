import React from 'react'
import { shallow, mount, render } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import Definition from './Definition'

test('snapshot test', () => {
  const component = shallow(<Definition handleGameStart={() => {}} />)
  const tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})

test('should not start game if player names are not valid', () => {
  const component = shallow(<Definition handleGameStart={() => {}} />)
  component.find('button').simulate('click')

  expect(component.find('.help.is-danger')).toHaveLength(2)
})

test('should show error message if player name is blank', () => {
  const component = shallow(<Definition handleGameStart={() => {}} />)
  component
    .find('input')
    .first()
    .simulate('change', { target: { value: '' } })

  expect(component.find('.help.is-danger').text()).toBeTruthy()
})

test('should show error message if players names are the same', () => {
  const component = shallow(<Definition handleGameStart={() => {}} />)
  component
    .find('input')
    .forEach(input => input.simulate('change', { target: { value: 'p1' } }))

  expect(component.find('.help.is-danger').text()).toBeTruthy()
})

test('should start game after start button is clicked if players names are valid', () => {
  const gameStartMock = jest.fn()
  const component = shallow(<Definition handleGameStart={gameStartMock} />)
  component
    .find('input')
    .forEach((input, index) =>
      input.simulate('change', { target: { value: `p${index}` } })
    )

  component.find('button').simulate('click')
  expect(gameStartMock).toHaveBeenCalled()
})
