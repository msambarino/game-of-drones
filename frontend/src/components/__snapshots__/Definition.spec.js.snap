// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="container"
>
  <section
    className="section is-small columns is-centered"
  >
    <div
      className="column is-half"
    >
      <h3
        className="title has-text-centered"
      >
        Enter Player's Names
      </h3>
      <div
        className="field"
      >
        <label
          className="label is-normal"
        >
          Player 1
        </label>
        <div
          className="field"
        >
          <div
            className="control"
          >
            <input
              className="input null"
              onChange={[Function]}
              placeholder="Player 1 name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="field"
      >
        <label
          className=" label is-normal"
        >
          Player 2
        </label>
        <div
          className="field"
        >
          <div
            className="control"
          >
            <input
              className="input null"
              onChange={[Function]}
              placeholder="Player 2 name"
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="control"
      >
        <button
          className="button is-primary is-fullwidth"
          onClick={[Function]}
        >
          Start
        </button>
      </div>
    </div>
  </section>
</div>
`;
